EIPS/eip-xxxx.md

```diff
@@ -0,0 +1,99 @@
+
+## Abstract
+
+Currently, the EVM provides basic storage functionalities for smart contracts, primarily in the form of mappings (key-value pairs). However, mappings are unordered, and there is no built-in support for iterating over them in a sorted order. This proposal suggests adding a new functionality to the EVM to support a sorted KV store, enabling smart contracts to:
+
+1. **Store key-value pairs in a persistent and ordered manner**.
+2. **Efficiently iterate over stored keys in ascending or descending order**.
+3. **Improve contract execution and state management for applications like rankings, order-books and lists**.
+
+This will reduce the need for external libraries or off-chain computations to sort and manage data, improving efficiency and reducing gas costs.
+
+## Motivation
+
+Many decentralized applications (dApps) rely on ordered data for various use cases, such as:
+
+- **Ranking systems** (e.g., leaderboards or voting systems).
+- **Efficient data storage** for applications needing quick access to the most recent or highest priority items (e.g., token voting, auctions).
+
+Currently, to implement such ordered key-value pair storage, smart contracts must implement complex sorting logic or rely on external services (e.g., off-chain indexing). These approaches are often gas-inefficient and require extra off-chain infrastructure. By natively supporting sorted KV stores, smart contracts can perform these tasks more efficiently and cost-effectively.
+
+## Specification
+
+This EIP introduces a **native Sorted Key-Value Store** in the EVM, accessed through a **precompiled contract**. The store supports lexicographically ordered key-value pairs with full CRUD functionality and cursor-based iteration.
+
+```solidity=
+interface ISortedKVStore {
+    /// Inserts or updates a key-value pair
+    function set(bytes calldata key, bytes calldata value) external;
+
+    /// Retrieves the value associated with a given key
+    function get(bytes calldata key) external view returns (bytes memory value);
+
+    /// Removes a key-value pair from the store
+    function remove(bytes calldata key) external;
+    
+    /// Initializes an iterator starting at the first key >= `prefix`
+    function begin(bytes calldata prefix) external view returns (bytes32 iterator);
+
+    /// Advances the iterator and returns the next key-value pair
+    function next(bytes32 iterator) external view returns (
+        bytes32 newIterator,
+        bytes memory key,
+        bytes memory value
+    );
+}
+```
+
+### Notes
+
+* **Lexicographical Sorting**: Keys are sorted using byte-wise lexicographical order.
+* **Gas Efficiency**: All operations are designed to be efficient. The underlying Ethereum database (e.g., LevelDB or RocksDB) natively supports key iteration and sorted access.
```

Comment:
One thing you have to note is that the keys are hashed when inserting an item in the trie. This means the sorted access doesn't come out of the box.

Reply:
Yup! A separate storage is probably needed since a kv-store may have keys of different sizes (even longer than 32 bytes)

Reply:
I just wonder how a "storage proof" would look like in that case? Since we definitely cant store a sorted kv store within a patricia trie

Reply:
keys are of fixed size in slit storage, doesn't matter how they are stored (hashed or not), you can iterate them as 32 byte numbers

Reply:
We can't mix the data we want to index with the contract storage data. As you mentioned, the keys in contract storage are often the result of a hash, so contract storage entries would be scattered randomly among our indexed data, making the whole approach ineffective. Therefore, a separate storage system is needed.

Reply:
> We can't mix the data we want to index with the contract storage data. As you mentioned, the keys in contract storage are often the result of a hash, so contract storage entries would be scattered randomly among our indexed data, making the whole approach ineffective. Therefore, a separate storage system is needed.

I don't think the alternative is workable. There is a reason why the keys are hashed in the trie. It's to avoid targeted attacks for making a part of the trie very deep. The tree which contains the storage should be roughly uniform. Not saying that the feature you are seeking is unworkable, just this solution might not pan out.

Reply:
@s1na Yes there has to be a way!